src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs

```diff
@@ -220,6 +220,12 @@ public object ExplicitDefaultValue
         }
 
         /// <summary>
+        /// Returns the default expression of the parameter,
+        /// or null if the parameter doesn't have a default expression.
+        /// </summary>
+        internal abstract BoundExpression ExplicitDefaultExpression { get; }
```

Comment:
The more I think about this approach (introducing this API so that DataFlowPass can get to the bound node to record what locals/fields are used), the more I don’t like it. Some of the negative points are:
- We have to bind default parameter values in local functions twice;
- This pattern is hard to extend to deal with attribute arguments.

I think that the approach that we have taken about making sure that declaration diagnostics for local functions gets to the right place in the right time is a good pattern to follow. To briefly summarize the approach:
- We added ```virtual void AddDeclarationDiagnostics(DiagnosticBag diagnostics)``` to symbol that, by default, stores the diagnostics to the global diagnostics bag on Compilation.
- We provided an override in LocalFunctionSymbol, which stores the diagnostics inside the symbol. After we finish binding local function declaration we add that diagnostics to the method body diagnostic.
- We added an override to LocalFunctionTypeParameterSymbol and SourceParameterSymbol that simply delegates the work to the containing symbol.
-  We changed all the places that were reporting declaration diagnostics to use the new method to store it. That affected default parameter binding, type parameter binding, attribute binding, etc.

Here is a gist of my proposal:
- Add a ```virtual void RecordUsageOfFieldsAndLocals(BoundExpression expression)``` to symbol that, by default, records usage of fields in the global compilation state.
- Provide an override for a LocalFunctionSymbol, which stores used fields and locals (both) in a cache within the symbol. When DataFlowPass visits local function declaration, content of that cache is merged into the DataFlowPass data structures.
- Provide and override on LocalFunctionTypeParameterSymbol and SourceParameterSymbol that simply delegates the work to the containing symbol.
- Call the API in SourceComplexParameterSymbol in ``` protected ConstantValue MakeDefaultExpression(DiagnosticBag diagnostics, Binder binder)``` after the line
```
var convertedExpression = binderForDefault.BindParameterDefaultValue(defaultSyntax, parameterType, diagnostics, out valueBeforeConversion);
```
passing ```convertedExpression``` as the argument. No other changes needed in SourceComplexParameterSymbol class.
- Call the API on the target of the attribute from ``` internal CSharpAttributeData GetAttribute(AttributeSyntax node, NamedTypeSymbol boundAttributeType, DiagnosticBag diagnostics)``` (in Binder) with ```boundAttribute``` as the argument, you might need to thread the target symbol to this API (it appears binder instance doesn’t always reflects that accurately). Note that ```boundAttribute``` will contain bound nodes for all arguments in the attribute application, visiting it should cover all of them at once. Also, note that it doesn’t look like we need to do anything related to SemanticModel here because the GetAttribute function is not called from SemanticModel (please do keep the SemanticModel tests though). No other changes are needed in binder_attributes.cs.

CC @agocke